In [96]:
!pip install -q -U google-genai

In [97]:
from google.colab import userdata

GEMINI_API_KEY = userdata.get("GEMINI_API_KEY")

print("API key loaded successfully!")

API key loaded successfully!


In [98]:
from google import genai

client = genai.Client(api_key=GEMINI_API_KEY)

In [99]:
response = client.models.generate_content(
    model="gemini-3.5-flash-lite",
    contents="Say hello in English, Russian and Kazakh."
)

print(response.text)

Here is "hello" in all three languages:

* **English:** Hello
* **Russian:** Здравствуйте (Zdravstvuyte - formal) / Привет (Privet - informal)
* **Қазақша (Kazakh):** Сәлеметсіз бе (Salemetsiz be - formal) / Сәлем (Salem - informal)


In [101]:
!python part0_tokenizers.py

PART 0 -- OpenAI tokenizer pair. No API key, no cost, no network.

SENTENCE
------------------------------------------------------------------
                         EN           RU           KK
o200k_base               12           18           21
  ÷ EN                1.00x        1.50x        1.75x
cl100k_base              12           35           58
  ÷ EN                1.00x        2.92x        4.83x

COMPLAINT
------------------------------------------------------------------
                         EN           RU           KK
o200k_base               59           80          118
  ÷ EN                1.00x        1.36x        2.00x
cl100k_base              59          146          265
  ÷ EN                1.00x        2.47x        4.49x

SYSTEM_PROMPT
------------------------------------------------------------------
                         EN           RU           KK
o200k_base               40           49           67
  ÷ EN                1.00x        1.23x        1

In [110]:
!python part1_offline.py

PART 1 -- offline size measures. No API key, no cost, no network.

SENTENCE
------------------------------------------------------------------
                         EN           RU           KK
chars                    69           75           72
bytes                    69          139          135
words                    11           11            9
bytes/EN              1.00x        2.01x        1.96x
bytes/char             1.00         1.85         1.88

COMPLAINT
------------------------------------------------------------------
                         EN           RU           KK
chars                   300          315          344
bytes                   300          576          640
words                    54           45           42
bytes/EN              1.00x        1.92x        2.13x
bytes/char             1.00         1.83         1.86

SYSTEM_PROMPT
------------------------------------------------------------------
                         EN           RU         

1. A Latin letter is 1 byte in UTF-8; Cyrillic is 2. Kazakh letters
   outside the Russian alphabet (ә, ғ, қ, ң, ө, ұ, ү, һ, і) are also 2.
   You can see that in the bytes/char row above.

2. How many TOKENS will the Russian
   version cost, as a multiple of the English one?

   RU / EN = 1.92×

3. for Kazakh:

   KK / EN = 2.13×

4. On what did you base the prediction -- characters, bytes, or words?

   I based the prediction on UTF-8 byte size.

5. Which one, if any, would a tokenizer have any reason to follow?

   A tokenizer does not have to follow characters, bytes, or words exactly.
   It works with learned token/subword units, so byte size can only be used
   as a rough estimate.

In [104]:
!pip install -q -U google-genai

In [105]:
!python part2_measure_gemini.py

counting tokens on gemini-3.5-flash-lite
  sentence       en=13  ru=18  kk=31
  complaint      en=60  ru=79  kk=149
  system_prompt  en=41  ru=52  kk=88
  request        en=109  ru=139  kk=245  (system + complaint preview; exact count replaces it with --call)

wrote measurements.json


In [106]:
%run part2_measure_gemini.py --call --model gemini-3.5-flash-lite

counting tokens on gemini-3.5-flash-lite
  sentence       en=13  ru=18  kk=31
  complaint      en=60  ru=79  kk=149
  system_prompt  en=41  ru=52  kk=88
  request        en=109  ru=139  kk=245  (system + complaint preview; exact count replaces it with --call)

answering the same complaint on gemini-3.5-flash-lite, in each language:

[en]
  --- answer ---
  I am sorry, but I cannot see any attached contract or statement, and I do not have access to your account details or documents from your message. 
  
  To help me explain the basis for any rate change and address your concern, please provide the specific terms and conditions or deposit agreement from your documents. Without the text of the provided documents, I cannot confirm the rate or the terms of your account.
  usage: 101 in, 81 visible out, 0 thinking, 81 billed out
  total tokens reported by API: 182

[ru]
  --- answer ---
  Извините, но в предоставленных документах (в тексте вашего запроса) нет информации об условиях изменени

In [108]:
%run part3_cost_gemini.py --requests-per-day 2000 --model gemini-3.5-flash-lite

prices from https://ai.google.dev/gemini-api/docs/pricing
checked 2026-09-21
tokens measured on gemini-3.5-flash-lite
answer length: measured in Part 2, including thinking tokens
NOTE: tables use paid-tier list prices for production planning. Eligible Free Tier usage can be $0.

ONE SUPPORT REQUEST -- tokens and paid-tier cost in US cents
------------------------------------------------------------------------------
                                     EN            RU            KK
input tokens                        101           131           237
output tokens                        81            82           113
gemini-3.5-flash-lite            0.0233        0.0244        0.0354
gemini-3.6-flash                 0.0379        0.0406        0.0602
gemini-3.8-flash                 0.0440        0.0467        0.0686

AT 2,000 REQUESTS/DAY -- annualized paid-tier USD/year
------------------------------------------------------------------------------
                                     